### PR TITLE
fix: use custom errors to reduce gas consumption

### PR DESCRIPTION
**Pull request by nmlinaric**
## Description
Use `revert Custom Errro()` instead of `require` across contracts to reduce gas consumption.

## Related Issue Or Context
Closes: #179 

## How Has This Been Tested? Testing details.
unit tests

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [x] I have added tests to cover my changes.
- [x] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: contracts/ERC20Safe.sol [hunk 1]
```diff
@@ -90,16 +92,10 @@ contract ERC20Safe {
...
-        }         
+        }
         require(tokenSize > 0, "ERC20: not a contract");
 
-        (bool success, bytes memory returndata) = address(token).call(data);
-        require(success, "ERC20: call failed");
-
-        if (returndata.length > 0) {
-
-            require(abi.decode(returndata, (bool)), "ERC20: operation did not succeed");
```
Comment (lastperson) [inline contracts/ERC20Safe.sol:101]: Logic that verifies the returned value to be true/false must remain.
Need to have a test for this case.

**Review by lastperson — CHANGES_REQUESTED** [review]

**Comment by nmlinaric** [comment]
NOTE: Truffle doesn't support decoding custom errors and can behave weird - this was the case with `ERC20Safe`.
Besides that we like more the consistent usage of `require` in `ERC20Safe` which can be properly tested.
Regarding gas consumption - it's more or less the same using `require` compared to `revert CustomError()`.

File: test/helpers.js [hunk 2]
```diff
@@ -361,6 +361,27 @@ const createPermissionlessGenericExecutionData = (
...
+    await promise;
+  } catch (error) {
+    const encoded = web3.eth.abi.encodeFunctionSignature(expectedErrorSignature);
+    const returnValue = Object.entries(error.data).filter(
+      (it) => it.length > 1
+    ).map(
+      (it) => it[1]
+    ).find(
+      (it )=> it != null && it.constructor.name === "Object" && "return" in it
+    ).return
```
Comment (lastperson) [inline test/helpers.js:377]:
```suggestion
      (it) => it != null && it.constructor.name === "Object" && "return" in it
    ).return;
```

**Review by lastperson — CHANGES_REQUESTED** [review]
Tiny style fix.

**Review by MakMuftic — APPROVED** [review]

**Review by stevenjla — APPROVED** [review]

**Review by lastperson — APPROVED** [review]

**Pull request merged**